pkg/resourceapply/apps.go

```diff
@@ -17,7 +20,31 @@ func ApplyStatefulSetWithControl(
 	required *appsv1.StatefulSet,
 	options ApplyOptions,
 ) (*appsv1.StatefulSet, bool, error) {
-	return ApplyGeneric[*appsv1.StatefulSet](ctx, control, recorder, required, options)
+	return ApplyGenericWithHandlers[*appsv1.StatefulSet](
+		ctx,
+		control,
+		recorder,
+		required,
+		options,
+		nil,
+		func(required *appsv1.StatefulSet, existing *appsv1.StatefulSet) (string, metav1.DeletionPropagation, error) {
+			if !equality.Semantic.DeepEqual(existing.Spec.Selector, required.Spec.Selector) {
+				existingPodLabels := existing.Spec.Template.Labels
+				requiredSelector, err := metav1.LabelSelectorAsSelector(required.Spec.Selector)
+				if err != nil {
+					return "", metav1.DeletePropagationBackground, fmt.Errorf("can't parse required StatefulSet selector: %w", err)
+				}
+
+				if !requiredSelector.Matches(labels.Set(existingPodLabels)) {
+					return "", metav1.DeletePropagationBackground, fmt.Errorf("required StatefulSet selector %s doesn't match existing Pod Labels set %v", required.Spec.Selector, existingPodLabels)
+				}
+
+				return "spec.selector is immutable", metav1.DeletePropagationOrphan, nil
```

Comment:
nit: returning empty string would give user the native error from sts or are you trying to compensate for it not being specific enough?

Reply:
Non empty recreate reason controls whether deletion happens and is being part of the log message associated with it.
https://github.com/scylladb/scylla-operator/blob/master/pkg/resourceapply/helpers.go#L321-L327

Reply:
I've mixed the lines, disregard this